### ISSUE TITLE
Apply some automatic fixes

### ISSUE THREAD

**Issue by ndmitchell** [ndmitchell/ghcid#15]
There are certain classes of errors where the error points immediately to the solution. Certain language extensions are easy to turn on and obvious when required, and for certain functions you know where they live and could add an import. I'm particularly interested in cases where currently I have to go to the top of the file, as that's annoying and breaks my flow, and is usually just boilerplate. (It goes without saying, this would have to be opt-in.) Some examples of errors we could catch:

```
Illegal `..' in record pattern
Use -XRecordWildCards to permit this
```

Or:

```
Not in scope: `when'
```

Or:

```
The import of `Data.List' is redundant
  except perhaps to import instances from `Data.List'
To import instances alone, use: import Data.List()
```

**Comment by ndmitchell** [comment]
One way to trigger automatic fixes would be to do something like save the file, add a ?, then save the file again. Editors could then automate that interaction if they wanted. We could have a flag `--auto=?` so people could change to something else.

**Comment by rvion** [comment]
I +1 for some `--auto` stuff flag

I think it makes sense for ghcid to {be aware of / handle} file modifications trigger since 
 - it already watch for *.hs changes and have the machinery for file watching
 - it might affects when to :reload things / when to trigger things / when not :reload stuff.
   for example, if a trigger return an exitcode of -1, ghcid could decide to consider not to reload the file (dumb exemple, just for illustration)
 - it's easier than configuring some buggy `https://atom.io/packages/save-commands` plugin specific to each editor / running manual shell commands (with platform specific tools) alongside ghcid

I personally use scripts to automate `hlint` + `apply-refact`  and `stylish-haskell` like 
```
for file in **/*.hs; do
	stylish-haskell -i {file} && hlint -v --refactor --refactor-options="-i" {file};
done
```

I would be happy to change it to 
ghcid {--on-change | --auto | --pre-reload | --pre...}="stylish-haskell -i {file} && hlint -v --refactor --refactor-options=\"-i\" {file};"

alos, a --post trigger command would be great to have (so we can ring bell, show system notification, etc)

**Comment by ndmitchell** [comment]
You can already run commands you want on a successful reload with the `--test` flag - e.g. `--test=:!stylish-haskell -i *.hs`. I'm wary about going too far down the watching and applying route - what you've got above is pretty simple, and wiring it into ghcid means it only applies for files currently loaded - not files that are not loaded due to importing an error.

**Comment by rvion** [comment]
indeed, and I completely forgot about ghci `:!`, thanks for pointing that out
